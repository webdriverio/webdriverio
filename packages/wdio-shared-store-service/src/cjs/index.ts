const esmModule = import('../index.js')

exports.setValue = async function (key: string, value: never) {
    const m = await esmModule
    return m.setValue(key, value)
}

exports.getValue = async function (key: string) {
    const m = await esmModule
    return m.getValue(key)
}

exports.setResourcePool = async function (key: string, value: never) {
    const m = await esmModule
    return m.setResourcePool(key, value)
}

exports.getValueFromPool = async function (key: string, options: never) {
    const m = await esmModule
    return m.getValueFromPool(key, options)
}

exports.addValueToPool = async function (key: string, value: never) {
    const m = await esmModule
    return m.addValueToPool(key, value)
}