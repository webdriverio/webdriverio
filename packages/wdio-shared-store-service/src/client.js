import { post } from 'got'
import logger from '@wdio/logger'

const log = logger('@wdio/shared-store-service')

let baseUrl
export const setPort = (port) => { baseUrl = `http://localhost:${port}` }

/**
 * make a request to the server to get a value from the store
 * @param   {string} key
 * @returns {*}
 */
export const getValue = async (key) => {
    const res = await post(`${baseUrl}/get`, { json: { key }, responseType: 'json' }).catch(errHandler)
    return (res && res.body) ? res.body.value : undefined
}

/**
 * make a request to the server to set a value to the store
 * @param {string}  key
 * @param {*}       value `store[key]` value (plain object)
 */
export const setValue = async (key, value) => {
    await post(`${baseUrl}/set`, { json: { key, value } }).catch(errHandler)
}

const errHandler = err => {
    log.warn(err.statusCode, err.statusMessage, err.url, err.body)
}
