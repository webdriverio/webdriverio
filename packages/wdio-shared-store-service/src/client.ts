import got, { Response } from 'got'
import logger from '@wdio/logger'

const log = logger('@wdio/shared-store-service')

let baseUrl: string
export const setPort = (port: string) => { baseUrl = `http://localhost:${port}` }

/**
 * make a request to the server to get a value from the store
 * @param   {string} key
 * @returns {*}
 */
export const getValue = async (key: string) => {
    const res = await got.post(`${baseUrl}/get`, { json: { key }, responseType: 'json' }).catch(errHandler)
    return (res && res.body) ? (res.body as WebdriverIO.JsonObject).value : undefined
}

/**
 * make a request to the server to set a value to the store
 * @param {string}  key
 * @param {*}       value `store[key]` value (plain object)
 */
export const setValue = async (key: string, value: WebdriverIO.JsonCompatible | WebdriverIO.JsonPrimitive) => {
    await got.post(`${baseUrl}/set`, { json: { key, value } }).catch(errHandler)
}

const errHandler = (err: Response<Error>) => {
    log.warn(err.statusCode, err.statusMessage, err.url, err.body)
}
