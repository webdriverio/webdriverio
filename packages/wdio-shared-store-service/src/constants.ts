export const CUSTOM_CAP = 'wdio:sharedStoreServicePort'
