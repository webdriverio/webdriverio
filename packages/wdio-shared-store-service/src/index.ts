import type { JsonPrimitive, JsonCompatible } from '@wdio/types'

import SharedStoreLauncher from './launcher'
import SharedStoreService from './service'

export default SharedStoreService
export const launcher = SharedStoreLauncher

export interface BrowserExtension {
    sharedStore: {
        get: (key: string) => JsonPrimitive | JsonCompatible;
        set: (key: string, value: JsonPrimitive | JsonCompatible) => void;
    }
}

declare global {
    namespace WebdriverIOAsync {
        interface Browser extends BrowserExtension { }
        interface MultiRemoteBrowser extends BrowserExtension { }
    }

    namespace WebdriverIOSync {
        interface Browser extends BrowserExtension { }
        interface MultiRemoteBrowser extends BrowserExtension { }
    }
}
