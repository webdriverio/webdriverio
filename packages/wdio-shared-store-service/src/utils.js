import { join as pathJoin } from 'path'
import { promisify } from 'util'
import { readFile as readFileCb, writeFile as writeFileCb, unlink  } from 'fs'

export const readFile = promisify(readFileCb)
export const writeFile = promisify(writeFileCb)
export const deleteFile = promisify(unlink)

export const getPidPath = (pid) => pathJoin(__dirname, `/${pid}.pid`)
