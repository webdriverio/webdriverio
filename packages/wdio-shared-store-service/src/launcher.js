import logger from '@wdio/logger'

import { writeFile, deleteFile, getPidPath } from './utils'

const log = logger('@wdio/shared-store-service')

let server = null

export default class SharedStoreLauncher {
    constructor() {
        // current process pid
        this.pidFile = getPidPath(process.pid)
    }

    async onPrepare () {
        server = require('./server').default
        const result = await server.startServer()

        log.info(`Started shared server on port ${result.port}`)
        await writeFile(this.pidFile, result.port.toString())
    }

    async onComplete () {
        await server.stopServer()
        await deleteFile(this.pidFile)
    }
}
