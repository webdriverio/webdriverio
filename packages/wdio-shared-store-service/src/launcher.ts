import logger from '@wdio/logger'

import { writeFile, deleteFile, getPidPath } from './utils'
import { setPort } from './client'

const log = logger('@wdio/shared-store-service')

let server: SharedStoreServer

export default class SharedStoreLauncher {
    pidFile: string
    constructor() {
        // current process pid
        this.pidFile = getPidPath(process.pid)
    }

    async onPrepare () {
        server = require('./server').default
        const result = await server.startServer()
        setPort(result.port)

        log.info(`Started shared server on port ${result.port}`)
        await writeFile(this.pidFile, result.port.toString())
    }

    async onComplete () {
        await server.stopServer()
        await deleteFile(this.pidFile)
    }
}
