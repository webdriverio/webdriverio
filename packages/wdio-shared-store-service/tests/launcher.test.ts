import { writeFile, deleteFile } from '../src/utils'
import SharedStoreLauncher from '../src/launcher'
import StoreServerType from '../src/server'
const StoreServer: typeof StoreServerType = require('../src/server').default

const { stopServer } = StoreServer

jest.mock('../src/server', () => ({
    default: {
        startServer: async () => ({ port: 3000 }),
        stopServer: jest.fn(),
    }
}))
jest.mock('../src/utils', () => ({
    writeFile: jest.fn(),
    deleteFile: jest.fn(),
    getPidPath: (pid: number) => pid,
}))

const storeLauncher = new SharedStoreLauncher()

describe('SharedStoreService', () => {
    it('onPrepare', async () => {
        await storeLauncher.onPrepare()
        expect(writeFile).toBeCalledWith(process.pid, '3000')
    })

    it('onComplete', async () => {
        await storeLauncher.onComplete()
        expect(stopServer).toBeCalled()
        expect(deleteFile).toBeCalledWith(process.pid)
    })

    afterEach(() => {
        (writeFile as jest.Mock).mockClear()
        ;(deleteFile as jest.Mock).mockClear()
        ;(stopServer as jest.Mock).mockClear()
    })
})
