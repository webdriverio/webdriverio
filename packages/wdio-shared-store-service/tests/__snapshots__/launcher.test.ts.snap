// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`SharedStoreService > should update capabilities correctly > using multiremote caps 1`] = `
{
  "browserA": {
    "capabilities": {
      "browserName": "chrome",
      "wdio:sharedStoreServicePort": 3000,
    },
  },
  "browserB": {
    "capabilities": {
      "browserName": "firefox",
      "wdio:sharedStoreServicePort": 3000,
    },
  },
}
`;

exports[`SharedStoreService > should update capabilities correctly > using parallel multiremote caps 1`] = `
[
  {
    "browserA": {
      "capabilities": {
        "browserName": "chrome",
        "wdio:sharedStoreServicePort": 3000,
      },
    },
    "browserB": {
      "capabilities": {
        "browserName": "firefox",
        "wdio:sharedStoreServicePort": 3000,
      },
    },
  },
]
`;

exports[`SharedStoreService > should update capabilities correctly > using standard caps 1`] = `
[
  {
    "browserName": "chrome",
    "wdio:sharedStoreServicePort": 3000,
  },
]
`;

exports[`SharedStoreService > should update capabilities correctly > using w3c caps 1`] = `
[
  {
    "alwaysMatch": {
      "browserName": "chrome",
      "wdio:sharedStoreServicePort": 3000,
    },
    "firstMatch": [],
  },
]
`;
